### DISCUSSION TITLE
Can this be used in Maui Hybrid ?

### DISCUSSION THREAD

**Discussion by Arslan007** [AdamEssenmacher/MemoryToolkit.Maui#2] [Q&A]
@AdamEssenmacher thanks for this wonderful toolkit. Just wondering if this can be used easily in Maui Hybrid.

Thanks!

**Comment by AdamEssenmacher** [comment]
Probably, though I haven't tried it myself. There are three distinct pieces of the toolkit that may or may not be useful or interesting to you depending on the application.

1) `GCCollectionMonitor` should be usable directly with anything .NET. It's actually pretty simple; it just creates weak references, forces a few GC runs, and then checks to see if the tracked objects are still alive. It would just be on you to use its API manually.

2) `GCCollectionBehavior` is specific to .NET MAUI and can be attached to any `VisualElement` including `BlazorWebView`. How useful this might be to you depends on how 'hybrid' your app actually is. For example, if your app has a single root `BlazorWebView` that is expected to live for the entirety of your app lifecycle, then monitoring for its collection would be pointless.

On the other hand, if you're using `BlazorWebView` intermixed with other MAUI components and/or across various MAUI views, ContentPages, and/or NavigationPages... then `GCCollectionBehavior` can monitor each `BlazorWebView` for garbage collection as it is unloaded. It doesn't know anything about Blazor, so it won't be able to monitor individual Blazor components for leaks. It would only let you know if the `BlazorWebView` host control itself is collected or not.

3) `AutoDisconnectBehvahior` is also specific to .NET MAUI and, just like `GCCollectionBehavior`, can be attached to any `VisualElement` including `BlazorWebView`. Also just like `GCCollectionBehavior`, it doesn't know anything about Blazor and (at best) will only be able to compartmentalize (and possibly prevent) leaks in the BlazorWebView itself.

I would be interested to know more about your specific use case. Are you experiencing memory issues in a hybrid app already? Or are you just looking to preemptively monitor and prevent potential issues?

**Comment by Arslan007** [comment]
I have Maui hybrid have which uses `BlazorWebView` on `MainPage.xaml` as standard and then there are different Razor Pages which  uses different controls from DexExpress and Yes I am experiencing memory issues as at times app uses go to 12 GB or so after little use of app.

**Comment by AdamEssenmacher** [comment]
> at times app uses go to 12 GB or so after little use of app

...😶 that's a lot!

It sounds like your app only has one `BlazorWebView`, which means the two behaviors in this toolkit (`AutoDisconnectBehvior` and `GCMonitorBehavior`) probably won't be useful. You might find some use for `GCCollectionMonitor`. In any case, it sounds like you've got some tough memory profiling ahead of you!

**Reply by Arslan007** [reply]
Thanks Adam!

The weird thing is this only happens for Web only as my app have Blazor Web, WinUI (Windows) & iOS and as per initial observation this is only affecting Web so wish me luck for memory leaks hunting :)
